### PR TITLE
JP-3470: Add ifu covariance scaling to extract1d

### LINKED ISSUE #8083: IFU 1d spectral extraction should account for covariance
_Issue [JP-3470](https://jira.stsci.edu/browse/JP-3470) was created on JIRA by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw):_

As discussed in detail by Law+23 ([https://ui.adsabs.harvard.edu/abs/2023AJ....166...45L/abstract,)](https://ui.adsabs.harvard.edu/abs/2023AJ....166...45L/abstract,) the MIRI and NIRSpec IFU data cubes have significant covariance between adjacent spatial elements.  The amount depends on the type of cube built and the wavelength, but can easily result in a factor of 1.5-2 (for MRS) difference in the actual noise of an extracted spectrum vs the pipeline-reported ERR values in the 1d spectrum.

This can be accounted for via a simple scaling model as described in Law+23 above (Fig 7 in particular, attached here). While this article focused on the scaling factors appropriate for uniform extended sources, a similar exercise can be performed in the opposite limit of point sources whose variance is dominated by Poisson noise instead.  Doing such a simulation of injecting an artificial PSF into a scene, adding poisson-dominated noise into the cal.fits file, it's possible to similarly compare the nominal uncertainty reported in the extracted 1d spectrum to the bootstrapped uncertainty from multiple realizations of the injected noise compared to the noiseless source.  These simulations likewise point to a factor ~2 for point sources as well; this is rather more complicated than the extended source case, as it involves an annular background subtraction as well.

In Ch3B for instance the factor increases from 1.6 at r=0.5 FWHM, to 2.0 at r=2 FWHM, to 2.1 at r=3 FWHM.  This is in keeping with the asymptotic trend observed by Law+.

This broad consistency suggests that most of the necessary improvement would be achieved simply by multiplying all errors extracted from MRS data cubes by a factor of X = 1.8 (mid-way between values of about 1.6 found by Law+ for uniform sources, and values of 2.0 for a similar analysis of point sources).  The uncertainty about this correction factor for a given real observation is likely comparable to the difference between wavebands and apertures, and thus a single multiplicative correction factor is (for now) most warranted.  While it may be possible in future to refine this correction at the 20% level with a more complicated algorithm, it would be helpful to address the factor of 1.8 now as this would be a trivial change.

Question for [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) what the preferred way of doing this would be.  I can see two practical options right now:

1) Add the factor of X directly into the Extract1D code for MIRI MRS.

2) Add a new optional parameter ifu_covar_scale to Extract1D, code default to 1.0, and have MRS parameter reference file set X to the necessary value.

(1) is simpler, but (2) is probably more in keeping with best practice.  While we don't anticipate using different values, it allows users to do so if they wish.  Likewise, it would allow for NIRSpec to provide a non-unity scaling factor too at some point if that proves useful.  (I suspect it will, but I don't know what the value is).

The only downside that I can see is that (2) would be a little harder to unwind if, in some future year, there's a better algorithm for this than simply multiplying by a constant, but I don't see this as particularly problematic.

Also tagging [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) [Bethan James](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bjames)

### PR DESCRIPTION
**Pull request by drlaw1558**
Resolves [JP-3470](https://jira.stsci.edu/browse/JP-3470) by allowing for a parameter to be passed into extract1d accounting for IFU cube covariance.

Closes #8083

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Review by jemorrison — APPROVED** [review]
Looks like a straight forward addition.

File: docs/jwst/extract_1d/description.rst [hunk 1]
```diff
@@ -297,7 +297,13 @@ is used to set the resampling value. The default value is 10.
...
 VAR_POISSON, VAR_RNOISE, and VAR_FLAT arrays.  As such, ``extract_1d`` only propagates this
-non-differentiated error term.  Note that while covariance is also extremely important for IFU data cubes
+non-differentiated error term.  Since covariance is also extremely important for IFU data cubes
+(see discussion by Law et al. 2023; AJ, 166, 45) the optional parameter ``ifu_covar_scale``
+will multiply all ERR arrays in the extracted spectra by a constant prefactor to account
+for this covariance.  As discussed by Law et al. 2023, this prefactor provides
+a reasonable first-order correction for the vast majority of use cases.  Values for the prefactor
+are provided in the ``extract_1d`` parameter reference files for MIRI and NIRSpec.
+
 (as the IFUs themselves are significantly undersampled) this term is not presently computed or taken
```
Comment (hbushouse) [inline docs/jwst/extract_1d/description.rst:307]: this existing sentence no longer seems to flow from the addition of new content above

**Review by hbushouse — COMMENTED** [review]
Looks good. Just one comment on the updated documentation.

**Comment by hbushouse** [comment]
regtest run started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1424

**Comment by hbushouse** [comment]
CI tests are failing with:
```
self = <LogRecord: stpipe, 20, /home/runner/work/jwst/jwst/jwst/extract_1d/ifu.py, 248, "Applying ERR covariance prefactor of %g">

    def getMessage(self):
        """
        Return the message for this LogRecord.
    
        Return the message for this LogRecord after merging any user-supplied
        arguments with the message.
        """
        msg = str(self.msg)
        if self.args:
>           msg = msg % self.args
E           TypeError: must be real number, not NoneType
```

File: docs/jwst/extract_1d/description.rst (hunk 1 above)
Comment (drlaw1558) [reply docs/jwst/extract_1d/description.rst]: Derp, fixed.

**Comment by drlaw1558** [comment]
> CI tests are failing with:
> 
> ```
> self = <LogRecord: stpipe, 20, /home/runner/work/jwst/jwst/jwst/extract_1d/ifu.py, 248, "Applying ERR covariance prefactor of %g">
> 
>     def getMessage(self):
>         """
>         Return the message for this LogRecord.
>     
>         Return the message for this LogRecord after merging any user-supplied
>         arguments with the message.
>         """
>         msg = str(self.msg)
>         if self.args:
> >           msg = msg % self.args
> E           TypeError: must be real number, not NoneType
> ```

Hm, I thought the default argument of 1.0 in the function call would be enough to deal with this, but I guess not.  Added a check for value defined prior to doing anything with it.

**Comment by nden** [comment]
I'd like to check the issue with the parameter default value not being picked up before we merge this, to verify it's not a larger issue.

**Comment by hbushouse** [comment]
Started another regtest run at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1454

And the CI tests seem to suddenly be working OK. Not sure why ...

**Comment by hbushouse** [comment]
Latest regtest run is clean. Only failures are unrelated (ref file updates).

**Review by hbushouse — APPROVED** [review]

**Comment by nden** [comment]
I tracked the problem with the arguments to a test
jwst/extract_1d/tests/test_ifu_image_ref.py: test_ifu_3d
which calls [extract.do_extract_1d](https://github.com/spacetelescope/jwst/blob/master/jwst/extract_1d/tests/test_ifu_image_ref.py#L40) directly without passing a value for the parameter and so it uses the default in the [do_extract1d function](https://github.com/spacetelescope/jwst/blob/6c643b4fec6cadf74970d9565803e296d6bf5363/jwst/extract_1d/extract.py#L2726) which is `None`.

So passing of parameters works as expected but if extract.do_extract_1d is to be called on its own I suggest we set a [default value](https://github.com/spacetelescope/jwst/blob/6c643b4fec6cadf74970d9565803e296d6bf5363/jwst/extract_1d/extract.py#L2742) in the signature as well. Also remove the condition [here](https://github.com/spacetelescope/jwst/pull/8457/files#diff-3c1b552b138d80a05dba93e36e2e116439914228c3eef588b1380c4f278ad043R248).

**Comment by hbushouse** [comment]
Alternatively, instead of removing the condition [here](https://github.com/spacetelescope/jwst/pull/8457/files#diff-3c1b552b138d80a05dba93e36e2e116439914228c3eef588b1380c4f278ad043R248), it could be modified to only trigger when `ifu_covar_scale` is not equal to 1.0. That way you avoid having the unnecessary message about it being applied when it really isn't having any effect on the data.

**Comment by drlaw1558** [comment]
I've changed that None default to 1.0, and changed the conditional to trigger when the value is not 1.0

**Comment by hbushouse** [comment]
Final updates look good to me. Merging.

**Pull request merged**